### PR TITLE
Server supports custom headers

### PR DESCRIPTION
**Pull request by Shadow243**
### Description

Based on a discussion from the [Stalwart GitHub issue https://github.com/stalwartlabs/mail-server/discussions/477](https://github.com/stalwartlabs/mail-server/discussions/477), it appears that Stalwart's IMAP server only indexes standard headers defined in RFCs and does not support searching custom headers (such as `X-Auto-Bcc`) by default. This design choice is intended to optimize server performance by maintaining a manageable index size, as most users typically search by common fields like recipients, subjects, or message contents.

### Key Points
- **Custom Header Search Not Supported:** Stalwart currently does not support searching for custom headers (e.g., `X-Auto-Bcc`) because these headers are not indexed.

A `server_supports_custom_headers` function is implemented to determine if the server supports searching custom headers. This function sends a test command to the server and inspects the response to check for support.

[Related Issue: https://github.com/cypht-org/cypht/issues/1149](https://github.com/cypht-org/cypht/issues/1149)

Screenshot 2024-08-09 at 03 35 41

### CODE REVIEW

**Comment by marclaporte** [comment]
@Shadow243 I am fine to do a workaround for just one server, but please add a link to https://github.com/stalwartlabs/mail-server/discussions/477 in the source code so we remove the workaround if ever Stalwart no longer needs it.

**Comment by Shadow243** [comment]
> @Shadow243 I am fine to do a workaround for just one server, but please add a link to [stalwartlabs/mail-server#477](https://github.com/stalwartlabs/mail-server/discussions/477) in the source code so we remove the workaround if ever Stalwart no longer needs it.

Done.

File: modules/imap/hm-imap.php [hunk 1]
```diff
@@ -1099,7 +1099,7 @@ public function search($target='ALL', $uids=false, $terms=array(), $esearch=arra
             if ($only_auto_bcc) {
                $fld .= ' HEADER X-Auto-Bcc cypht';
             }
-            if (!mb_strstr($this->server, 'yahoo') && $exclude_auto_bcc) {
+            if ($this->server_supports_custom_headers() && !mb_strstr($this->server, 'yahoo') && $exclude_auto_bcc) {
```
Comment (kroky) [inline modules/imap/hm-imap.php:1102]: Note that this adds one more IMAP query to every search query which will have a performance impact. I suggest you reverse the order of checks here, so it is only called when needed - e.g. $exclude_auto_css && ! mb_strstr && $this->server_supports_custom_headers...

Comment (Shadow243) [reply modules/imap/hm-imap.php]: Done

**Comment by marclaporte** [comment]
Potential regression: https://github.com/cypht-org/cypht/issues/1195

File: modules/imap/hm-imap.php [hunk 2]
```diff
@@ -1099,7 +1099,7 @@ public function search($target='ALL', $uids=false, $terms=array(), $esearch=arra
             if ($only_auto_bcc) {
```
Comment (Shadow243) [inline modules/imap/hm-imap.php:1099]: I think the same conditions must have been here too

**Comment by marclaporte** [comment]
A necessary follow up: https://github.com/cypht-org/cypht/pull/1651#issuecomment-3249394383

**Pull request merged**